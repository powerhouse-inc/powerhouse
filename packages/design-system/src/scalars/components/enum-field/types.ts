import type { IconName } from "#powerhouse";

export interface RadioGroupProps {
  options?: {
    value: string;
    label: string;
    description?: string;
    disabled?: boolean;
  }[];
  onChange?: (value: string) => void;
}

export interface SelectOption {
  icon?: IconName | React.ComponentType<{ className?: string }>;
  value: string;
  label: string;
  disabled?: boolean;
}

export interface SelectBaseProps {
  options?: SelectOption[];
  favoriteOptions?: SelectOption[];
  placeholder?: string;
  multiple?: boolean;
  searchable?: boolean;
  onChange?: (value: string | string[]) => void;
  contentClassName?: string;
  contentAlign?: "start" | "end" | "center";
}

export type SelectProps =
  | (SelectBaseProps & {
      selectionIcon?: "auto";
      selectionIconPosition?: "left";
    })
  | (SelectBaseProps & {
      selectionIcon: "checkmark";
      selectionIconPosition?: "left" | "right";
    });
