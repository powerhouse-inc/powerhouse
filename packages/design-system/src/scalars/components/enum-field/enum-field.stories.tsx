import { Icon, type IconName } from "#powerhouse";
import type { Meta, StoryObj } from "@storybook/react";

import { withForm } from "../../lib/decorators.js";
import {
  getDefaultArgTypes,
  getValidationArgTypes,
  StorybookControlCategory,
} from "../../lib/storybook-arg-types.js";
import { EnumField } from "./enum-field.js";

const meta: Meta<typeof EnumField> = {
  title: "Document Engineering/Scalars/Enum Field",
  component: EnumField,
  decorators: [
    withForm,
    (Story) => (
      <div style={{ maxWidth: "280px", margin: "1rem auto 0" }}>
        <Story />
      </div>
    ),
  ],
  parameters: {
    layout: "padded",
  },
  tags: ["autodocs"],
  argTypes: {
    ...getDefaultArgTypes(),

    variant: {
      control: "radio",
      options: ["auto", "RadioGroup", "Select"],
      description:
        "Enum field variant. " +
        "auto: uses the most appropriate variant based on the number of options " +
        "(less than 6 options -> RadioGroup, 6 options or more -> Select). " +
        "RadioGroup: displays options in a group of radio buttons. " +
        "Select: displays options in a dropdown menu.",
      table: {
        type: { summary: '"auto" | "RadioGroup" | "Select"' },
        defaultValue: { summary: "auto" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    options: {
      control: "object",
      description:
        "Array of options with label, value, icon, description and disabled",
      table: {
        type: {
          summary:
            "Array<{ label: string; value: string; icon?: IconName | React.ComponentType<{ className?: string }>; description?: string; disabled?: boolean; }>",
        },
        defaultValue: { summary: "[]" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    multiple: {
      control: "boolean",
      description: "Whether multiple options can be selected",
      table: {
        type: { summary: "boolean" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    selectionIcon: {
      control: "radio",
      options: ["auto", "checkmark"],
      description:
        "Selection icon to show in the options. " +
        "auto: Show a Radio for Single Select and a Checkbox for Multi Select. " +
        "checkmark: Show a checkmark icon for Single and Multi Select. ",
      table: {
        type: { summary: '"auto" | "checkmark"' },
        defaultValue: { summary: '"auto"' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    selectionIconPosition: {
      control: "radio",
      options: ["left", "right"],
      description:
        'Selection icon position in options. Only apply if "selectionIcon" is "checkmark".',
      table: {
        type: { summary: '"left" | "right"' },
        defaultValue: { summary: '"left"' },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
      if: { arg: "selectionIcon", eq: "checkmark" },
    },

    searchable: {
      control: "boolean",
      description: "Whether to enable search functionality",
      table: {
        type: { summary: "boolean" },
        category: StorybookControlCategory.COMPONENT_SPECIFIC,
      },
    },

    ...getValidationArgTypes(),
  },
  args: {
    name: "enum-field",
    errors: [],
    warnings: [],
  },
} satisfies Meta<typeof EnumField>;

export default meta;

type Story = StoryObj<typeof EnumField>;

const IconComponent = (
  name: IconName,
): React.ComponentType<{ className?: string }> => {
  return ({ className }) => (
    <Icon name={name} size={16} className={className} />
  );
};

const defaultOptions = [
  { value: "Briefcase", label: "Briefcase" },
  { value: "Drive", label: "Drive" },
  { value: "Globe", label: "Globe" },
  { value: "Settings", label: "Settings" },
];

const defaultOptionsWithIcon = [
  {
    value: "Briefcase",
    label: "Briefcase",
    icon: IconComponent("Briefcase"),
  },
  {
    value: "Drive",
    label: "Drive",
    icon: IconComponent("Drive"),
  },
  {
    value: "Globe",
    label: "Globe",
    icon: IconComponent("Globe"),
  },
  {
    value: "Settings",
    label: "Settings",
    icon: IconComponent("Settings"),
  },
];

type StoryProps = {
  variant: "auto" | "RadioGroup" | "Select";
  [key: string]: any;
};

export const Default: Story = {
  args: {
    label: "Select an option",
    options: defaultOptions,
    placeholder: "Choose from the list",
  },
};

export const WithDescription: Story = {
  args: {
    label: "Select an option",
    description: "Choose your preferred option from the list",
    variant: "RadioGroup",
    options: defaultOptions.map((opt, index) => ({
      ...opt,
      description:
        index === 2 ? `Description for ${opt.label} option` : undefined,
    })),
    placeholder: "Choose from the list",
  } as StoryProps,
};

export const Required: Story = {
  args: {
    label: "Required field",
    variant: "RadioGroup",
    options: defaultOptions,
    placeholder: "Choose from the list",
    required: true,
  } as StoryProps,
};

export const WithDefaultValue: Story = {
  args: {
    label: "Preset selection",
    variant: "RadioGroup",
    options: defaultOptions,
    placeholder: "Choose from the list",
    defaultValue: "Drive",
  } as StoryProps,
};

export const Disabled: Story = {
  args: {
    label: "Disabled field",
    variant: "RadioGroup",
    options: defaultOptions,
    placeholder: "Choose from the list",
    value: "Drive",
    disabled: true,
  } as StoryProps,
};

// Validation states
export const WithError: Story = {
  args: {
    label: "With error",
    variant: "RadioGroup",
    options: defaultOptions,
    placeholder: "Choose from the list",
    value: "Drive",
    errors: ["Please select a different option"],
  } as StoryProps,
};

export const WithWarning: Story = {
  args: {
    label: "With warning",
    variant: "RadioGroup",
    options: defaultOptions,
    placeholder: "Choose from the list",
    value: "Drive",
    warnings: ["This option may be deprecated soon"],
  } as StoryProps,
};

// Variant examples
export const Select: Story = {
  args: {
    label: "Select variant",
    variant: "Select",
    options: defaultOptions,
    placeholder: "Select an option",
  },
};

export const SelectWithDescription: Story = {
  args: {
    label: "Select variant with descriptions",
    description: "Choose your preferred option",
    variant: "Select",
    options: defaultOptions,
    placeholder: "Select an option",
  },
};

export const MultiSelect: Story = {
  args: {
    label: "Multi Select variant",
    variant: "Select",
    options: defaultOptions,
    placeholder: "Select options",
    multiple: true,
  },
};

export const SearchableSelect: Story = {
  args: {
    label: "Searchable select variant",
    variant: "Select",
    options: defaultOptions,
    placeholder: "Search and select an option",
    searchable: true,
  },
};

export const SelectWithCheckmark: Story = {
  args: {
    label: "With checkmark",
    description: "Shows a checkmark for selected options",
    variant: "Select",
    options: defaultOptions,
    selectionIcon: "checkmark",
    placeholder: "Select an icon name",
  },
};

export const SelectWithIcon: Story = {
  args: {
    label: "Favorite icon",
    description: "Choose your favorite icon",
    variant: "Select",
    options: defaultOptionsWithIcon,
    selectionIcon: "checkmark",
    selectionIconPosition: "right",
    placeholder: "Select an icon",
  },
};
