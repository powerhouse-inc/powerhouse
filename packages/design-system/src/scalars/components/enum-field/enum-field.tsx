import React from "react";
import {
  RadioGroupField,
  type RadioGroupFieldProps,
} from "../fragments/radio-group-field/radio-group-field.js";
import {
  SelectField,
  type SelectFieldProps,
} from "../fragments/select-field/select-field.js";

export type EnumFieldProps =
  | ({
      variant?: "auto";
    } & (RadioGroupFieldProps | SelectFieldProps))
  | ({
      variant: "RadioGroup";
    } & RadioGroupFieldProps)
  | ({
      variant: "Select";
    } & SelectFieldProps);

export const EnumField = React.forwardRef<
  HTMLDivElement | HTMLButtonElement,
  EnumFieldProps
>(({ variant = "auto", options = [], ...props }, ref) => {
  const radio = (
    <RadioGroupField
      options={options}
      {...(props as RadioGroupFieldProps)}
      ref={ref as React.ForwardedRef<HTMLDivElement>}
    />
  );
  const select = (
    <SelectField
      options={options}
      {...(props as SelectFieldProps)}
      ref={ref as React.ForwardedRef<HTMLButtonElement>}
    />
  );

  switch (variant) {
    case "RadioGroup":
      return radio;
    case "Select":
      return select;
    case "auto":
      return options.length < 6 ? radio : select;
  }
});

EnumField.displayName = "EnumField";
