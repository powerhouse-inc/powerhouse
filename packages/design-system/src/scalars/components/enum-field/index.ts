export { EnumField, type EnumFieldProps } from "./enum-field.js";
