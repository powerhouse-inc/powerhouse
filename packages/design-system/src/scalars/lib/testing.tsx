import { render, type RenderResult } from "@testing-library/react";
import { Form } from "../components/index.js";

export const renderWithForm = (
  children: React.ReactNode,
  onSubmit?: () => void,
): RenderResult => {
  return render(children, {
    wrapper: ({ children }) => (
      <Form onSubmit={onSubmit ?? (() => null)}>{children}</Form>
    ),
  });
};
