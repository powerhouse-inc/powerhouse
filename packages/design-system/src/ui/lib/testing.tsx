import { Form } from "@powerhousedao/design-system";
import type { RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";

export const renderWithForm = (
  children: React.ReactNode,
  onSubmit?: () => void,
): RenderResult => {
  return render(children, {
    wrapper: ({ children }) => (
      <Form onSubmit={onSubmit ?? (() => null)}>{children}</Form>
    ),
  });
};
