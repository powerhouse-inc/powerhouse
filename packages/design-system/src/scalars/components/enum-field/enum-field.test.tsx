import { screen, waitFor } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import { describe, expect, it, vi } from "vitest";
import { renderWithForm } from "../../lib/testing.js";
import { EnumField } from "./enum-field.js";

describe("EnumField Component", () => {
  const defaultProps = {
    name: "enum",
    variant: "RadioGroup" as const,
    options: [
      { value: "option1", label: "Option 1" },
      { value: "option2", label: "Option 2" },
      { value: "option3", label: "Option 3" },
    ],
  };
  window.HTMLElement.prototype.scrollIntoView = vi.fn();
  window.Element.prototype.scrollTo = vi.fn();

  it("should match snapshot", () => {
    const { asFragment } = renderWithForm(<EnumField {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should match snapshot with Select variant", () => {
    const { asFragment } = renderWithForm(
      <EnumField {...defaultProps} variant="Select" />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render label when provided", () => {
    renderWithForm(<EnumField {...defaultProps} label="Test Label" />);
    expect(screen.getByText("Test Label")).toBeInTheDocument();
  });

  it("should render description when provided", () => {
    renderWithForm(
      <EnumField
        {...defaultProps}
        variant="Select"
        description="Test description"
      />,
    );
    expect(screen.getByText("Test description")).toBeInTheDocument();
  });

  it("should render warning messages when provided", () => {
    renderWithForm(
      <EnumField {...defaultProps} warnings={["Warning message"]} />,
    );
    expect(screen.getByText("Warning message")).toBeInTheDocument();
  });

  it("should render error messages when provided", async () => {
    renderWithForm(<EnumField {...defaultProps} errors={["Error message"]} />);
    await waitFor(() =>
      expect(screen.getByText("Error message")).toBeInTheDocument(),
    );
  });

  it("should handle value changes in RadioGroup variant", async () => {
    const user = userEvent.setup();
    const onChange = vi.fn();
    renderWithForm(
      <EnumField {...defaultProps} value="option1" onChange={onChange} />,
    );

    const radio = screen.getByLabelText("Option 2");
    await user.click(radio);

    expect(onChange).toHaveBeenCalledWith("option2");
  });

  it("should handle value changes in Select variant", async () => {
    const user = userEvent.setup();
    const onChange = vi.fn();
    renderWithForm(
      <EnumField
        {...defaultProps}
        variant="Select"
        multiple
        value={["option1"]}
        onChange={onChange}
      />,
    );

    const select = screen.getByRole("combobox");
    await user.click(select);
    await user.click(screen.getByText("Option 2"));
    expect(onChange).toHaveBeenCalledWith(["option1", "option2"]);
  });

  it("should show required indicator when required prop is true", () => {
    renderWithForm(<EnumField {...defaultProps} label="Test Label" required />);
    expect(screen.getByText("*")).toBeInTheDocument();
  });

  it("should disable the field when disabled prop is true", () => {
    renderWithForm(<EnumField {...defaultProps} variant="Select" disabled />);
    const select = screen.getByRole("combobox");
    expect(select).toBeDisabled();
  });

  it("should default to Select variant when no variant provided and has 7 options", () => {
    const propsWithSevenOptions = {
      name: "enum",
      options: [
        { value: "option1", label: "Option 1" },
        { value: "option2", label: "Option 2" },
        { value: "option3", label: "Option 3" },
        { value: "option4", label: "Option 4" },
        { value: "option5", label: "Option 5" },
        { value: "option6", label: "Option 6" },
        { value: "option7", label: "Option 7" },
      ],
    };
    renderWithForm(<EnumField {...propsWithSevenOptions} />);
    // Verify it renders as a Select
    expect(screen.getByRole("combobox")).toBeInTheDocument();
  });
});
