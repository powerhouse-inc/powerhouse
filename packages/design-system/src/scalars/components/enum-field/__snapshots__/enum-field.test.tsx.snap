// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EnumField Component > should match snapshot 1`] = `
<DocumentFragment>
  <form
    id=":r0:"
    novalidate=""
  >
    <div
      aria-invalid="false"
      aria-label="Radio group"
      class="flex flex-col gap-2"
      dir="ltr"
      id=":r1:-radio-group"
      role="radiogroup"
      style="outline-color: none; outline-style: none; outline-width: initial;"
      tabindex="0"
    >
      <div
        class="flex items-center gap-2"
        role="presentation"
      >
        <button
          aria-checked="false"
          aria-disabled="false"
          aria-invalid="false"
          class="group peer aspect-square size-4 rounded-full border border-gray-800 dark:border-gray-400 hover:border-gray-900 dark:hover:border-gray-50"
          data-radix-collection-item=""
          data-state="unchecked"
          id=":r1:-radio-0-option1"
          role="radio"
          tabindex="-1"
          type="button"
          value="option1"
        />
        <input
          aria-hidden="true"
          name="enum"
          style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px; width: 0px; height: 0px;"
          tabindex="-1"
          type="radio"
          value="option1"
        />
        <label
          class="inline-flex items-center text-sm font-medium leading-[22px] text-gray-900 dark:text-gray-400 group-hover:text-gray-900 dark:group-hover:text-slate-50 cursor-pointer peer-hover:text-gray-900 dark:peer-hover:text-gray-50"
          for=":r1:-radio-0-option1"
        >
          Option 1
        </label>
      </div>
      <div
        class="flex items-center gap-2"
        role="presentation"
      >
        <button
          aria-checked="false"
          aria-disabled="false"
          aria-invalid="false"
          class="group peer aspect-square size-4 rounded-full border border-gray-800 dark:border-gray-400 hover:border-gray-900 dark:hover:border-gray-50"
          data-radix-collection-item=""
          data-state="unchecked"
          id=":r1:-radio-1-option2"
          role="radio"
          tabindex="-1"
          type="button"
          value="option2"
        />
        <input
          aria-hidden="true"
          name="enum"
          style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px; width: 0px; height: 0px;"
          tabindex="-1"
          type="radio"
          value="option2"
        />
        <label
          class="inline-flex items-center text-sm font-medium leading-[22px] text-gray-900 dark:text-gray-400 group-hover:text-gray-900 dark:group-hover:text-slate-50 cursor-pointer peer-hover:text-gray-900 dark:peer-hover:text-gray-50"
          for=":r1:-radio-1-option2"
        >
          Option 2
        </label>
      </div>
      <div
        class="flex items-center gap-2"
        role="presentation"
      >
        <button
          aria-checked="false"
          aria-disabled="false"
          aria-invalid="false"
          class="group peer aspect-square size-4 rounded-full border border-gray-800 dark:border-gray-400 hover:border-gray-900 dark:hover:border-gray-50"
          data-radix-collection-item=""
          data-state="unchecked"
          id=":r1:-radio-2-option3"
          role="radio"
          tabindex="-1"
          type="button"
          value="option3"
        />
        <input
          aria-hidden="true"
          name="enum"
          style="transform: translateX(-100%); position: absolute; pointer-events: none; opacity: 0; margin: 0px; width: 0px; height: 0px;"
          tabindex="-1"
          type="radio"
          value="option3"
        />
        <label
          class="inline-flex items-center text-sm font-medium leading-[22px] text-gray-900 dark:text-gray-400 group-hover:text-gray-900 dark:group-hover:text-slate-50 cursor-pointer peer-hover:text-gray-900 dark:peer-hover:text-gray-50"
          for=":r1:-radio-2-option3"
        >
          Option 3
        </label>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`EnumField Component > should match snapshot with Select variant 1`] = `
<DocumentFragment>
  <form
    id=":r9:"
    novalidate=""
  >
    <div
      class="flex flex-col gap-2"
    >
      <button
        aria-controls="radix-:rb:"
        aria-expanded="false"
        aria-haspopup="dialog"
        aria-invalid="false"
        aria-label="Select"
        class="gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none disabled:pointer-events-none [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 flex h-9 w-full items-center justify-between px-3 py-2 dark:border-charcoal-700 dark:bg-charcoal-900 rounded-md border border-gray-300 bg-white hover:border-gray-300 hover:bg-gray-100 dark:hover:border-charcoal-700 dark:hover:bg-charcoal-800 dark:focus:ring-charcoal-300 focus:outline-none focus:ring-1 focus:ring-gray-900 focus:ring-offset-0 dark:focus-visible:ring-charcoal-300 focus-visible:ring-1 focus-visible:ring-gray-900 focus-visible:ring-offset-0"
        data-state="closed"
        id=":ra:-select"
        name="enum"
        role="combobox"
        type="button"
      >
        <div
          class="mx-auto flex w-full items-center justify-end"
        >
          <svg
            class="cursor-pointer text-gray-700 dark:text-gray-400"
            fill="none"
            stroke="currentcolor"
            style="width: 16px; height: 16px;"
            viewBox="0 0 24 24"
          >
            <path
              d="M6 9L12 15L18 9"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
            />
          </svg>
        </div>
      </button>
    </div>
  </form>
</DocumentFragment>
`;
